lhotse/features/base.py

```diff
@@ -0,0 +1,411 @@
+
+    An example of usage:
+
+    @register_extractor
+    class MyFeatureExtractor:
+        ...
+
+    :param cls: A type (class) that is being registered.
+    :return: Registered type.
+    """
+    FEATURE_EXTRACTORS[cls.name] = cls
+    return cls
+
+
+class TorchaudioFeatureExtractor(FeatureExtractor):
+    """Common abstract base class for all torchaudio based feature extractors."""
+    feature_fn = None
+
+    def extract(self, samples: np.ndarray, sampling_rate: int) -> np.ndarray:
+        params = asdict(self.config)
+        params.update({
+            "sample_frequency": sampling_rate,
+            "snip_edges": False
+        })
+        params['frame_shift'] *= 1000.0
+        params['frame_length'] *= 1000.0
+        if not isinstance(samples, torch.Tensor):
+            samples = torch.from_numpy(samples)
+        features = self.feature_fn(samples, **params)
+        return features.numpy()
+
+    @property
+    def frame_shift(self) -> Seconds:
+        return self.config.frame_shift
+
+
+@dataclass(order=True)
+class Features:
+    """
+    Represents features extracted for some particular time range in a given recording and channel.
+    It contains metadata about how it's stored: storage_type describes "how to read it", for now
+    it supports numpy arrays serialized with np.save, as well as arrays compressed with lilcom;
+    storage_path is the path to the file on the local filesystem.
+    """
+    recording_id: str
+    channel_id: int
+    start: Seconds
+    duration: Seconds
+
+    # Useful information about the features - their type (fbank, mfcc) and shape
+    type: str
+    num_frames: int
+    num_features: int
+    sampling_rate: int
+
+    # Parameters related to storage - they define how to load the feature matrix.
+    storage_type: str  # e.g. 'lilcom', 'numpy'
+    storage_path: str
+
+    @property
+    def end(self) -> Seconds:
+        return self.start + self.duration
+
+    @property
+    def frame_shift(self) -> Seconds:
+        return round(self.duration / self.num_frames, ndigits=3)
+
+    def load(
+            self,
+            root_dir: Optional[Pathlike] = None,
+            start: Seconds = 0.0,
+            duration: Optional[Seconds] = None,
+    ) -> np.ndarray:
+        # Load the features from the storage
+        storage_path = self.storage_path if root_dir is None else Path(root_dir) / self.storage_path
+        if self.storage_type == 'lilcom':
+            with open(storage_path, 'rb') as f:
+                features = lilcom.decompress(f.read())
+        elif self.storage_type == 'numpy':
+            features = np.load(storage_path, allow_pickle=False)
+        else:
+            raise ValueError(f"Unknown storage_type: {self.storage_type}")
+
+        # In case the caller requested only a subset of features, trim them
+
+        # Left trim
+        if not isclose(start, self.start):
+            frames_to_trim = round((start - self.start) / self.frame_shift)
```

Comment:
may not give informative error message if start < self.start?

Reply:
Good catch